### PR TITLE
[Fiber] Compute the Host Diff During Reconciliation

### PR DESCRIPTION
**Pull request by sebmarkbage**
This uses prepareUpdate to compute the diff to apply. If there is no work to do, we don't schedule any updates at all for the commit phase.

Unfortunately wrapper components have custom logic that gets applied in the commit phase so for now we need to always schedule an update for those.

The payload has an array of only the prop key and values that need to update.

This makes the commit phase in the Sierpinski Triangle demo go from 2.5-3ms to <1.5ms. The overall time probably increases but at least now we won't risk dropping frames as we're updating low-pri content.

A neat side-effect of this is that we no longer depend on any "contextual information" like host context or root instance in the commit phase.

TODO 1: ~~This doesn't correctly apply the diff when the `is` property switches from custom component mode to built-in component mode. I'll just need to deal with that when deleting values.~~ Errr... This is not true. We can't switch between modes. I just added some unused code at some point because I thought we could.

~~TODO 2: A more fundamental problem is that we always need to update the pointer to the "current" fiber atm because of how the lazy event listener thing works. I could probably change it to store the props object instead of going through the fiber.~~

~~TODO 3: Flow infinite loops. (Updating to 0.37.1 doesn't help. https://github.com/facebook/react/pull/8608)~~ Fixed using workaround.

### CODE REVIEW

File: src/renderers/shared/shared/event/EventPluginHub.js [hunk 1]
```diff
@@ -127,6 +127,10 @@ var EventPluginHub = {
     // live here; needs to be moved to a better place soon
     if (typeof inst.tag === 'number') {
       const props = inst.memoizedProps;
+      if (!props) {
+        // Work in progress.
```
Comment (sebmarkbage) [inline src/renderers/shared/shared/event/EventPluginHub.js]: This is part of TODO 2 because if we have no event listeners, then we don't update the fiber pointer. If we don't update the Fiber pointer, then the new one becomes work in progress. If we try to read from work in progress, this might be null. If it completes, it might have event listeners on it that are not yet committed.

Therefore this isn't a great solution.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 2]
```diff
@@ -690,17 +655,164 @@ var ReactDOMFiberComponent = {
...
+        var lastHtml = lastProp ? lastProp[HTML] : undefined;
+        if (nextHtml) {
+          if (lastHtml) {
+            if (lastHtml !== nextHtml) {
+              (updatePayload = updatePayload || []).push(propKey, '' + nextHtml);
+            }
+          } else {
+            (updatePayload = updatePayload || []).push(propKey, '' + nextHtml);
+          }
+        } else {
```
Comment (gaearon) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: This should probably be (from #8652):

```js
        if (nextHtml != null) {
          if (lastHtml !== nextHtml) {
              (updatePayload = updatePayload || []).push(propKey, '' + nextHtml);
          }
        } else {
          // TODO: ...
```

**Comment by gaearon** [comment]
(I used GH interface for resolving the conflict.)

**Comment by sebmarkbage** [comment]
I already rebased locally.

**Comment by gaearon** [comment]
:-( I should've asked.

**Comment by gaearon** [comment]
@sebmarkbage 

Did you use a time machine writing these commits? 😄 

To the next person utterly confused by GitHub's display of this PR, ba11c3ba335f8c533a54d66006bcf3ea109d6dd1 depends on 6ae87cfb2868c0658deea79d641ab6b92a028d74 but for some reason was committed a day earlier, and thus appears before it in the list.

screen shot 2017-01-11 at 23 59 26

**Comment by sebmarkbage** [comment]
Like politics, rebasing is all about rewriting history. In chronological order or not.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 3]
```diff
@@ -671,17 +629,161 @@ var ReactDOMFiberComponent = {
...
+         !lastProps.hasOwnProperty(propKey) ||
+         lastProps[propKey] == null) {
+        continue;
+      }
+      if (propKey === STYLE) {
+        var lastStyle = lastProps[propKey];
+        for (styleName in lastStyle) {
+          if (lastStyle.hasOwnProperty(styleName)) {
+            if (!styleUpdates) {
+              styleUpdates = {};
```
Comment (gaearon) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Could we pack style updates in the same array with special keys or would that be too complicated?

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 4]
```diff
@@ -671,17 +629,161 @@ var ReactDOMFiberComponent = {
...
+            typeof nextProp === 'string' || typeof nextProp === 'number'
+            )) {
+          (updatePayload = updatePayload || []).push(propKey, '' + nextProp);
+        }
+      } else if (propKey === SUPPRESS_CONTENT_EDITABLE_WARNING) {
+        // Noop
+      } else if (registrationNameModules.hasOwnProperty(propKey)) {
+        if (nextProp) {
+          // We eagerly listen to this even though we haven't committed yet.
+          ensureListeningTo(rootContainerElement, propKey);
```
Comment (gaearon) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Why do we listen here instead of the commit?

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 5]
```diff
@@ -433,18 +357,57 @@ function updateDOMProperties(
...
+}
+
+function updateDOMProperties(
+  domElement : Element,
+  updatePayload : Array<any>,
+  wasCustomComponentTag : boolean,
+  isCustomComponentTag : boolean,
+) : void {
+  // TODO: Handle wasCustomComponentTag
+  for (var i = 0; i < updatePayload.length; i+=2) {
```
Comment (gaearon) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Style nit: `i += 2`

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 6]
```diff
@@ -668,17 +636,160 @@ var ReactDOMFiberComponent = {
...
+          if (lastStyle.hasOwnProperty(styleName)) {
+            if (!styleUpdates) {
+              styleUpdates = {};
+            }
+            styleUpdates[styleName] = '';
+          }
+        }
+      } else if (propKey === DANGEROUSLY_SET_INNER_HTML ||
+                 propKey === CHILDREN) {
+        // TODO: Clear innerHTML. This is currently broken in Fiber because we are
```
Comment (gaearon) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: I think @acdlite said this was solved.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 7]
```diff
@@ -668,17 +636,160 @@ var ReactDOMFiberComponent = {
...
+        // too late to clear everything at this point because new children have
+        // already been inserted.
+      } else if (propKey === SUPPRESS_CONTENT_EDITABLE_WARNING) {
+        // Noop
+      } else if (registrationNameModules.hasOwnProperty(propKey)) {
+        // This is a special case. If any listener updates we need to ensure
+        // that the "current" fiber pointer gets updated so we need a commit
+        // to update this element.
```
Comment (gaearon) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: This depends on `precacheFiber` call in `ReactDOMFiber.js`, right? It's a bit non-obvious that an empty array here is needed to trigger the path there even though `updateProperties` in this file won't do anything. I guess doesn’t matter if we’ll unify them soon.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 8]
```diff
@@ -668,17 +636,160 @@ var ReactDOMFiberComponent = {
...
+      } else {
+        // For all other deleted properties we add it to the queue. We use
+        // the whitelist in the commit phase instead.
+        (updatePayload = updatePayload || []).push(propKey, null);
+      }
+    }
+    for (propKey in nextProps) {
+      var nextProp = nextProps[propKey];
+      var lastProp =
+        lastProps != null ? lastProps[propKey] : undefined;
```
Comment (gaearon) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: I don't think `lastProps` can be null here. You can probably read directly now that mount and update is separate.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 9]
```diff
@@ -668,17 +636,160 @@ var ReactDOMFiberComponent = {
...
+        if (!updatePayload) {
+          updatePayload = [];
+        }
+      } else {
+        // For all other deleted properties we add it to the queue. We use
+        // the whitelist in the commit phase instead.
+        (updatePayload = updatePayload || []).push(propKey, null);
+      }
+    }
+    for (propKey in nextProps) {
```
Comment (gaearon) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: While unrelated to this PR, I think the most common path is `lastProps` and `nextProps` with the same prop keys in the same order. I wonder if we could optimize this as a faster case with just one loop, and only do the second loop if we find a mismatch during the first one.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 10]
```diff
@@ -668,17 +636,160 @@ var ReactDOMFiberComponent = {
...
+            if (nextProp.hasOwnProperty(styleName) &&
+                lastProp[styleName] !== nextProp[styleName]) {
+              if (!styleUpdates) {
+                styleUpdates = {};
+              }
+              styleUpdates[styleName] = nextProp[styleName];
+            }
+          }
+        } else {
+          // Relies on `updateStylesByID` not mutating `styleUpdates`.
```
Comment (gaearon) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Stray comment.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 11]
```diff
@@ -668,17 +636,160 @@ var ReactDOMFiberComponent = {
...
+          if (!styleUpdates) {
```
Comment (gaearon) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: How can `styleUpdates` be truthy here? It could only get created if `lastStyle` existed and we wanted to delete some styles. But in this case we would get into `if (lastProp)` branch instead of this one. Unless I’m missing something.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 12]
```diff
@@ -668,17 +636,160 @@ var ReactDOMFiberComponent = {
...
+          updatePayload = [];
+        }
+      } else {
+        // For any other property we always add it to the queue and then we
+        // filter it out using the whitelist during the commit.
+        (updatePayload = updatePayload || []).push(propKey, nextProp);
+      }
+    }
+    if (styleUpdates) {
+      (updatePayload = updatePayload || []).push(STYLE, styleUpdates);
```
Comment (gaearon) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Is it still necessary to delay it until the end? Why? This structure makes the above style code a little hard to follow, especially the case where you have to reassign `styleUpdates` right after pushing it to the payload so that it doesn't get skipped when it's falsy.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js (hunk 3 above)
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Interestingly RN did this by packing all fields into one payload object. Caused a bunch of weird artifacts that made diffing harder to optimize.

We should reevaluate the actual algorithms here but for now I just wanted compatibility with CSSPropertyOperations since it is shared with Stack.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js (hunk 4 above)
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: One of the things this fixes is that `getRootHostContainer` is no longer needed in the commit phase if we do it here. As we know, `getRootHostContainer` is not safe to use during commits since we may not have the right context.

There's also no need to do this in the commit phase so might as well do it during the time sliced parts.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js (hunk 7 above)
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Also note that there is an open TODO items related to this.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js (hunk 9 above)
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: That assumes that you can iterate both interleaved which `for...in` doesn't let you do. It would require something like `Object.keys` or a generator with `for...in` to be really fast. Not too worried about this though. Something we can micro-optimize easily if needed.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js (hunk 8 above)
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Nice catch. I wish Flow would warn about this.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js (hunk 11 above)
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: The code in this branch is also wrong since it'll push a `null`. I'll delete.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js (hunk 12 above)
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Donno. I just did a line-by-line translation.

Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: If I don't put it here, I need it in all the places where I could create `styleUpdates`. Because there may not be a STYLE field in the nextProps but might be one in lastProps. I'm not sure that's better.

**Comment by sebmarkbage** [comment]
Ok, this I fixed the outstanding TODO by storing the "current" props on the DOM node and in the React Native instance cache. Not too happy about how this all came together but not really sure what we should do about the event system.

This is nice because it minimizes work during commit and initial mount but still relies on "diffing" event listeners during updates to figure out if we should update at all.

A plausible alternative model would be to traverse the "current" tree while processing the event, or even process events as part of reconciliation which does that traversal anyway.

File: src/renderers/shared/fiber/ReactFiberCompleteWork.js [hunk 13]
```diff
@@ -200,8 +201,12 @@ module.exports = function<T, P, I, TI, PI, C, CX>(
...
           const instance : I = workInProgress.stateNode;
           const currentHostContext = getHostContext();
-          if (prepareUpdate(instance, type, oldProps, newProps, currentHostContext)) {
-            // This returns true if there was something to update.
+          const updatePayload = prepareUpdate(instance, type, oldProps, newProps, rootContainerInstance, currentHostContext);
+          // TODO: Type this specific to this type of component.
+          workInProgress.updateQueue = (updatePayload : any);
+          // If the update payload indicates that there is a change or if there
+          // is a new ref we mark this as an update.
+          if (updatePayload || current.ref !== workInProgress.ref) {
```
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberCompleteWork.js]: Refs shouldn't schedule an `Update` effect anymore, since we added the `Ref` effect

**Comment by hellonewbie** [comment]
should I now delete the project?

11.01.2017 22:22 "Dan Abramov" <notifications@github.com> napisał(a):

File: scripts/fiber/tests-failing.txt [hunk 14]
```diff
@@ -37,6 +37,9 @@ src/renderers/dom/shared/__tests__/ReactRenderDocument-test.js
 * should throw on full document render w/ no markup
 * supports findDOMNode on full-page components
 
+src/renderers/dom/shared/wrappers/__tests__/ReactDOMInput-test.js
+* should control a value in reentrant events
```
Comment (gaearon) [inline scripts/fiber/tests-failing.txt]: I missed why this is failing.

Comment (sebmarkbage) [inline scripts/fiber/tests-failing.txt]: Ah yes. I missed this too. This is probably either a bug or behavior change regarding event timing.

Comment (sebmarkbage) [inline scripts/fiber/tests-failing.txt]: Fixed: https://github.com/facebook/react/pull/8839
